src/state_machine.zig

```diff
@@ -1599,6 +1898,86 @@ pub fn StateMachineType(
             return self.forest.grooves.posted.get(pending_timestamp);
         }
 
+        fn transfer_update_pending_status(
+            self: *StateMachine,
+            timestamp: u64,
+            status: TransferPendingStatus,
+        ) void {
+            assert(timestamp != 0);
+            assert(status != .none and status != .pending);
+
+            self.forest.grooves.transfers.indexes.pending_status.remove(&.{
+                .timestamp = timestamp,
+                .field = @intFromEnum(TransferPendingStatus.pending),
+            });
+            self.forest.grooves.transfers.indexes.pending_status.put(&.{
+                .timestamp = timestamp,
+                .field = @intFromEnum(status),
+            });
+        }
+
+        fn execute_expire_pending_transfers(self: *StateMachine, timestamp: u64) usize {
+            assert(self.scan_result_count <= constants.batch_max.create_transfers);
+
+            if (self.scan_result_count == 0) return 0;
+            defer self.scan_result_count = 0;
+
+            const grooves = &self.forest.grooves;
+            const transfers: []const Transfer = std.mem.bytesAsSlice(
+                Transfer,
+                self.scan_buffer[0 .. self.scan_result_count * @sizeOf(Transfer)],
+            );
+
+            log.debug("expire_pending_transfers: len={}", .{transfers.len});
+
+            for (transfers) |expired| {
+                assert(expired.flags.pending);
+                assert(expired.timeout > 0);
+
+                const expires_at = expired.timestamp + expired.timeout_ns();
+                assert(expires_at <= timestamp);
+
+                if (global_constants.verify) {
+                    assert(grooves.posted.get(expired.timestamp) == null);
+                }
+
+                const dr_account = grooves.accounts.get(
+                    expired.debit_account_id,
+                ).?;
+                assert(dr_account.debits_pending >= expired.amount);
+
+                const cr_account = grooves.accounts.get(
+                    expired.credit_account_id,
+                ).?;
+                assert(cr_account.credits_pending >= expired.amount);
+
+                var dr_account_new = dr_account.*;
+                var cr_account_new = cr_account.*;
+                dr_account_new.debits_pending -= expired.amount;
+                cr_account_new.credits_pending -= expired.amount;
+
+                grooves.accounts.update(.{ .old = dr_account, .new = &dr_account_new });
+                grooves.accounts.update(.{ .old = cr_account, .new = &cr_account_new });
+
+                // Expire the transfer.
+                grooves.posted.insert(&.{
+                    .timestamp = expired.timestamp,
+                    .fulfillment = .expired,
+                });
```

Comment:
Could the `grooves.posted.insert` be moved into `self.transfer_update_pending_status`?